File: examples/conduit/frontend/appkit/Services/APIClient.swift
// Conduit Example App -- AppKit API Client
// URLSession-based HTTP client for the Conduit REST API on macOS.

import Foundation

// MARK: - Models

struct User: Codable {
    let username: String
    let email: String
    let token: String
    let bio: String?
    let image: String?
}

struct Profile: Codable {
    let username: String
    let bio: String?
    let image: String?
    var following: Bool
}

struct Article: Codable {
    let slug: String
    let title: String
    let description: String
    let body: String
    let tagList: [String]
    let createdAt: String
    let updatedAt: String
    var favorited: Bool
    var favoritesCount: Int
    var author: Profile
}

struct Comment: Codable {
    let id: String
    let body: String
    let createdAt: String
    let author: Profile
}

struct UserResponse: Codable { let user: User }
struct ProfileResponse: Codable { let profile: Profile }
struct ArticleResponse: Codable { let article: Article }
struct ArticlesResponse: Codable { let articles: [Article]; let articlesCount: Int }
struct CommentResponse: Codable { let comment: Comment }
struct CommentsResponse: Codable { let comments: [Comment] }
struct TagsResponse: Codable { let tags: [String] }
struct ErrorBody: Codable { let body: [String] }
struct ErrorResponse: Codable { let errors: ErrorBody }

// MARK: - API Client

enum APIClientError: LocalizedError {
    case httpError(Int, String)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .httpError(_, let message): return message
        case .networkError(let error): return error.localizedDescription
        }
    }
}

class APIClient {
    static let shared = APIClient()

    private let baseURL = "http://localhost:3000"
    private let session = URLSession.shared
    private let decoder = JSONDecoder()

    var token: String?
    var currentUser: User?

    var isAuthenticated: Bool { token != nil }

    private func request<T: Decodable>(_ method: String, path: String, body: Any? = nil) async throws -> T {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw APIClientError.httpError(0, "Invalid URL")
        }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = token {
            req.setValue("Token \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await session.data(for: req)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIClientError.httpError(0, "Invalid response")
        }

        if httpResponse.statusCode >= 400 {
            if let errorRes = try? decoder.decode(ErrorResponse.self, from: data) {
                throw APIClientError.httpError(httpResponse.statusCode, errorRes.errors.body.joined(separator: ", "))
            }
            throw APIClientError.httpError(httpResponse.statusCode, "HTTP \(httpResponse.statusCode)")
        }

        return try decoder.decode(T.self, from: data)
    }

    // Auth
    func login(email: String, password: String) async throws -> User {
        let body: [String: Any] = ["user": ["email": email, "password": password]]
        let res: UserResponse = try await request("POST", path: "/api/users/login", body: body)
        token = res.user.token
        currentUser = res.user
        return res.user
    }

    func register(username: String, email: String, password: String) async throws -> User {
        let body: [String: Any] = ["user": ["username": username, "email": email, "password": password]]
        let res: UserResponse = try await request("POST", path: "/api/users", body: body)
        token = res.user.token
        currentUser = res.user
        return res.user
    }

    func logout() { token = nil; currentUser = nil }

    // Articles
    func getArticles() async throws -> [Article] {
        let res: ArticlesResponse = try await request("GET", path: "/api/articles")
        return res.articles
    }

    func getArticle(slug: String) async throws -> Article {
        let res: ArticleResponse = try await request("GET", path: "/api/articles/\(slug)")
        return res.article
    }

    func deleteArticle(slug: String) async throws {
        let _: [String: String] = try await request("DELETE", path: "/api/articles/\(slug)")
    }

    // Comments
    func getComments(slug: String) async throws -> [Comment] {
        let res: CommentsResponse = try await request("GET", path: "/api/articles/\(slug)/comments")
        return res.comments
    }

    // Social
    func follow(username: String) async throws -> Profile {
        let res: ProfileResponse = try await request("POST", path: "/api/profiles/\(username)/follow")
        return res.profile
    }

    func unfollow(username: String) async throws -> Profile {
        let res: ProfileResponse = try await request("DELETE", path: "/api/profiles/\(username)/follow")
        return res.profile
    }

    func favorite(slug: String) async throws -> Article {
        let res: ArticleResponse = try await request("POST", path: "/api/articles/\(slug)/favorite")
        return res.article
    }

    func unfavorite(slug: String) async throws -> Article {
        let res: ArticleResponse = try await request("DELETE", path: "/api/articles/\(slug)/favorite")
        return res.article
    }

    // Profile
    func getProfile(username: String) async throws -> Profile {
        let res: ProfileResponse = try await request("GET", path: "/api/profiles/\(username)")
        return res.profile
    }
}

File: examples/conduit/frontend/shared/api-client.ts
// Conduit Example App — Shared API Client
// HTTP fetch wrapper for the Conduit REST API.

import type {
  UserResponse, ProfileResponse, ArticleResponse,
  ArticlesResponse, CommentResponse, CommentsResponse,
  TagsResponse, ErrorResponse,
} from './types.js';

export class ConduitAPI {
  private baseUrl: string;
  private token: string | null = null;

  constructor(baseUrl = 'http://localhost:3000') {
    this.baseUrl = baseUrl;
  }

  setToken(token: string | null) {
    this.token = token;
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const headers: Record<string, string> = { 'Content-Type': 'application/json' };
    if (this.token) {
      headers['Authorization'] = `Token ${this.token}`;
    }

    const response = await fetch(`${this.baseUrl}${path}`, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    const data = await response.json();

    if (!response.ok) {
      const err = data as ErrorResponse;
      throw new Error(err.errors?.body?.join(', ') || `HTTP ${response.status}`);
    }

    return data as T;
  }

  // Auth
  async register(username: string, email: string, password: string): Promise<UserResponse> {
    return this.request('POST', '/api/users', { user: { username, email, password } });
  }

  async login(email: string, password: string): Promise<UserResponse> {
    return this.request('POST', '/api/users/login', { user: { email, password } });
  }

  // Profile
  async updateProfile(bio?: string, image?: string): Promise<UserResponse> {
    return this.request('PUT', '/api/user', { user: { bio, image } });
  }

  async getProfile(username: string): Promise<ProfileResponse> {
    return this.request('GET', `/api/profiles/${username}`);
  }

  // Articles
  async createArticle(title: string, description: string, body: string, tagList?: string[]): Promise<ArticleResponse> {
    return this.request('POST', '/api/articles', { article: { title, description, body, tagList } });
  }

  async deleteArticle(slug: string): Promise<void> {
    return this.request('DELETE', `/api/articles/${slug}`);
  }

  async getArticles(): Promise<ArticlesResponse> {
    return this.request('GET', '/api/articles');
  }

  // Comments
  async createComment(slug: string, body: string): Promise<CommentResponse> {
    return this.request('POST', `/api/articles/${slug}/comments`, { comment: { body } });
  }

  async deleteComment(slug: string, commentId: string): Promise<void> {
    return this.request('DELETE', `/api/articles/${slug}/comments/${commentId}`);
  }

  async getComments(slug: string): Promise<CommentsResponse> {
    return this.request('GET', `/api/articles/${slug}/comments`);
  }

  // Social
  async follow(username: string): Promise<ProfileResponse> {
    return this.request('POST', `/api/profiles/${username}/follow`);
  }

  async unfollow(username: string): Promise<ProfileResponse> {
    return this.request('DELETE', `/api/profiles/${username}/follow`);
  }

  async favorite(slug: string): Promise<ArticleResponse> {
    return this.request('POST', `/api/articles/${slug}/favorite`);
  }

  async unfavorite(slug: string): Promise<ArticleResponse> {
    return this.request('DELETE', `/api/articles/${slug}/favorite`);
  }

  // Tags
  async getTags(): Promise<TagsResponse> {
    return this.request('GET', '/api/tags');
  }

  // Health
  async health(): Promise<{ status: string; concepts: number; syncs: number }> {
    return this.request('GET', '/api/health');
  }
}
